zenoh-flow/src/runtime/dataflow/instance/runners/mod.rs

```diff
@@ -39,27 +45,207 @@ pub enum RunAction {
+    pub(crate) inputs_callbacks_abort_handle: Option<AbortHandle>,
+}
+
+impl Runner {
+    pub(crate) fn new(
+        node: Option<Arc<dyn Node>>,
+        inputs_callbacks: Vec<CallbackInput>,
+        outputs_callbacks: Vec<CallbackOutput>,
+    ) -> Self {
+        Self {
+            node,
+            inputs_callbacks,
+            outputs_callbacks,
+            run_loop_handle: None,
+            run_loop_abort_handle: None,
+            outputs_callbacks_handle: None,
+            outputs_callbacks_abort_handle: None,
+            inputs_callbacks_handle: None,
+            inputs_callbacks_abort_handle: None,
+        }
+    }
+
+    /// TODO(J-Loudet) Improve documentation.
+    ///
+    /// First, `start` is idempotent: no matter how many times this method is called, if the node is
+    /// running then nothing happens.
+    ///
+    /// This method creates as many `AbortHandle` as required to run what the user provided: (i) an
+    /// iteration, (ii) input_callbacks, (iii) output_callbacks.
+    ///
+    /// Considering all are present, we spawn a task for each and call their respective methods in
+    /// an infinite loop.
+    ///
+    ///
+    /// # Issues
+    ///
+    /// FIXME Assuming there are several callbacks for the same kind (i.e. input or output), if one
+    ///       of these callbacks never finishes, then all the others will starve. For instance, if
+    ///       an input callbacks sleeps (std::thread::sleep) for 10 seconds, then the other inputs
+    ///       won’t have access to the CPU for these 10 seconds.
+    ///       A possible fix would be to spawn a thread per callback. What’s the impact?
+    pub(crate) fn start(&mut self) {
+        if self.is_running() {
+            log::warn!("Called `start` while node is ALREADY running. Returning.");
+            return;
+        }
+
+        if let Some(node) = &self.node {
+            let node = Arc::clone(node);
+            let run_loop = async move {
+                let mut instant: Instant;
+                loop {
+                    instant = Instant::now();
+                    if let Err(e) = node.iteration().await {
+                        log::error!("Iteration error: {:?}", e);
+                        return e;
+                    }
+
+                    log::trace!("iteration took: {}ms", instant.elapsed().as_millis());
+
+                    async_std::task::yield_now().await;
+                }
+            };
+
+            let (abort_handle, abort_registration) = AbortHandle::new_pair();
+            let handle = async_std::task::spawn(Abortable::new(run_loop, abort_registration));
+
+            self.run_loop_handle = Some(handle);
+            self.run_loop_abort_handle = Some(abort_handle);
+        }
+
+        if !self.outputs_callbacks.is_empty() {
+            // TODO How is that clone "safe"? The underlying channels are "job stealing" so only the
+            // last copy actually gets the data?
```

Comment:
It is not the last that receive the data, but the one that is ready on the `recv_async().await`